options-service/src/options/options.repository.ts

```diff
@@ -0,0 +1,83 @@
+import * as C from "../config";
+import {
+  Options,
+  GetOptionsByEntityIdReq,
+  UpdateUserOptionsReq,
+  FavoriteCommanderReq,
+} from "../interfaces/protoc/proto/optionsService";
+import { OptionsModel } from "../models/options.model";
+import { logger } from "../logger";
+
+export class OptionsRepository {
+  async getOptionsByEntityId(
+    getOptionsByEntityIdRequest: GetOptionsByEntityIdReq
+  ): Promise<Options> {
+    const { entityId } = getOptionsByEntityIdRequest;
+    const setOnInsert = {
+      $setOnInsert: C.defaultOptions,
+    };
+    const document = await OptionsModel.findOneAndUpdate(
+      { entityId },
+      setOnInsert,
+      {
+        upsert: true,
+        new: true,
+      }
+    );
+    return document;
+  }
+
+  async updateUserOptions(
+    updateUserOptionsRequest: UpdateUserOptionsReq
+  ): Promise<Options> {
+    const { entityId, ...updateQuery } = updateUserOptionsRequest;
+    const updatedDocument = await OptionsModel.findOneAndUpdate(
+      { entityId },
+      updateQuery,
+      {
+        new: true,
+      }
+    );
+    if(!updatedDocument) return OptionsModel.create({ ...C.defaultOptions, ...updateUserOptionsRequest });
```

Comment:
1. **$set** called if found (ignoring $setOnInsert) 
2. **$setOnInsert** inserts a new record (and then executes $set). So you need to get rid of duplicate fields (otherwise it will throw an error)

```suggestion
    const subtractFields = (obj1, obj2) => Object.fromEntries(Object.entries(obj1).filter(([key]) => !(key in obj2)));
    const updatedDocument = await OptionsModel.findOneAndUpdate(
      { entityId },
      { $set: {updateQuery}, $setOnInsert: subtractFields(C.defaultOptions, updateQuery) },
      { new: true, upsert: true }
    );
    if(!updatedDocument) return OptionsModel.create({ ...C.defaultOptions, ...updateUserOptionsRequest });
```